font/src/ft/mod.rs

```diff
@@ -293,19 +294,22 @@ impl FreeTypeRasterizer {
         };
 
         // Reuse the font_key, since changing it can break library users
-        let font_key = self.face_from_pattern(&base_font, font_key).and_then(|pattern| {
-            pattern.map(Ok).unwrap_or_else(|| Err(Error::MissingFont(desc.to_owned())))
-        })?;
+        let font_key =
+            self.face_from_pattern(&base_pattern, base_hash, font_key).and_then(|pattern| {
+                pattern.map(Ok).unwrap_or_else(|| Err(Error::MissingFont(desc.to_owned())))
+            })?;
 
         // Coverage for fallback fonts
         let coverage = CharSet::new();
         let empty_charset = CharSet::new();
         // Load fallback list
-        let list: Vec<FallbackFont> = font_iter
-            .map(|font| {
-                let charset = font.get_charset().unwrap_or(&empty_charset);
+        let list: Vec<FallbackFont> = pattern_iter
+            .map(|fallback_pattern| {
+                let charset = fallback_pattern.get_charset().unwrap_or(&empty_charset);
+                let pattern = base_pattern.render_prepare(config, fallback_pattern);
+                let hash = FontHash::new(hash, pattern.hash());
                 let _ = coverage.merge(&charset);
-                FallbackFont::new(font.to_owned())
+                FallbackFont::new(pattern, hash)
```

Comment:
We've discussed on IRC that all we'd need to add to the hash would be the font size. What is the reason to hash the two different patterns here instead? Is it to include other settings that might be different and relevant, if so, what would those be?

Reply:
I don't want to use u64 hashes from rust, since XOR of 2 u32 hashes from Fontconfig should be already enough for our needs, I'm a bit afraid of collisions tbh, since we're not doing a "good" thing here. We can just use 2 strings(path + style + pixelsize) and hash by them, however I'm not sure about the performance of it.

Reply:
A u32 has about 4,294,967,295 different possible values. Assuming an even distribution of font hashes across this, it seems like collisions would be pretty unlikely?

It seems like we'd basically half the precision by trying to combine two different hashes that are potentially in conflict, but that should still leave us with plenty of headroom.

The most problematic thing would be if the input hashes were not evenly distributed. XORing the values should be okay-ish, assuming the two hashes are not identical. If it is possible that hashes could be identical, that would cause big problems since we'd get a lot of `0` hashes. It seems like a lot of people recommend `m * hash(a) + hash(b)` over XOR.

Reply:
where `m` is a prime number, I'd assume? As you can see those bitshifts are here not as decorations :), so I'm trying to prevent us from such thing.